dart/common/detail/CompositeData.h

```diff
@@ -0,0 +1,354 @@
+{
+public:
+
+  /// Forwarding constructor
+  template <typename... Args>
+  CompositeData(Args&&... args)
+    : CloneableMap<MapType>(std::forward<Args>(args)...)
+  {
+    // Do nothing
+  }
+
+  virtual ~CompositeData() = default;
+
+  /// Create (or replace) a piece of data in this
+  template <class AspectT, typename... Args>
+  typename GetData<AspectT>::Type& create(Args&&... args)
+  {
+    using Data = typename GetData<AspectT>::Type;
+    using AspectType = typename GetAspect<AspectT>::Type;
+    using DataType = typename GetData<Aspect>::Type;
+
+    std::unique_ptr<DataType>& data = this->mMap[typeid(AspectType)]
+        = make_unique<Data>(std::forward<Args>(args)...);
+    return static_cast<Data&>(*data);
+  }
+
+  template <class AspectT>
+  typename GetData<AspectT>::Type* get()
+  {
+    using Data = typename GetData<AspectT>::Type;
+    using AspectType = typename GetAspect<AspectT>::Type;
+
+    typename MapType::iterator it = this->mMap.find(typeid(AspectType));
+    if(this->mMap.end() == it)
+      return nullptr;
+
+    return static_cast<Data*>(it->second.get());
+  }
+
+  template <class AspectT>
+  const typename GetData<AspectT>::Type* get() const
+  {
+    return const_cast<CompositeData<MapType, GetData>*>(this)->get<AspectT>();
+  }
+
+  template <class AspectT, typename... Args>
+  typename GetData<AspectT>::Type& getOrCreate(Args&&... args)
+  {
+    using Data = typename GetData<AspectT>::Type;
+    using AspectType = typename GetAspect<AspectT>::Type;
+
+    auto& it = this->mMap.insert(
+          std::make_pair<std::type_index, std::unique_ptr<Data>>(
+            typeid(AspectType), nullptr));
+
+    const bool exists = !it.second;
+    if(!exists)
+      it.first = make_unique<Data>(std::forward<Args>(args)...);
+
+    return static_cast<Data&>(*it.first);
+  }
+
+};
+
+//==============================================================================
+using CompositeState = CompositeData<CompositeStateMap, GetState>;
+using CompositeProperties = CompositeData<CompositePropertiesMap, GetProperties>;
+
+//==============================================================================
+template <class CompositeType, template<class> class GetData, typename... Aspects>
+class ComposeData
+{
+public:
+
+  ComposeData() = default;
+
+  ComposeData(const CompositeType&)
+  {
+    // Do nothing
+  }
+
+  virtual ~ComposeData() = default;
+
+  void setFrom(const CompositeType&)
+  {
+    // Do nothing
+  }
+
+protected:
+
+  void _addData(CompositeType&) const
+  {
+    // Do nothing
+  }
+};
+
+//==============================================================================
+template <class CompositeType, template<class> class GetData, class AspectT,
+          typename... Remainder>
+struct ComposeData<CompositeType, GetData, AspectT, Remainder...> :
+    public GetData<AspectT>::Type,
+    public ComposeData<CompositeType, GetData, Remainder...>
+{
+public:
+
+  enum Delegate_t { Delegate };
```

Comment:
Honestly, it's because I can't think of a better way to distinguish the type name from the value name for enumeration tags. I'm open to alternative naming conventions.